sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -521,38 +524,138 @@ class Analyzer(
    */
   object ResolveSortReferences extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-      case s @ Sort(ordering, global, p @ Project(projectList, child))
-          if !s.resolved && p.resolved =>
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+      case s @ Sort(_, _, child) if !s.resolved && child.resolved =>
+        val (newOrdering, missingResolvableAttrs) =
+          collectResolvableMissingAttrs(s.order, plans = mutable.Queue(child))
 
-        // If this rule was not a no-op, return the transformed plan, otherwise return the original.
-        if (missing.nonEmpty) {
-          // Add missing attributes and then project them away after the sort.
-          Project(p.output,
-            Sort(newOrdering, global,
-              Project(projectList ++ missing, child)))
-        } else {
-          logDebug(s"Failed to find $missing in ${p.output.mkString(", ")}")
+        if (missingResolvableAttrs.isEmpty) {
+          val unresolvableAttrs = s.order.filterNot(_.resolved)
+          logDebug(s"Failed to find $unresolvableAttrs in ${child.output.mkString(", ")}")
           s // Nothing we can do here. Return original plan.
         }
+        else {
+          // Transform the whole tree in post-order. Add into the self's outputSet
+          //   all the children attributes that are part of missingResolvableAttrs.
+          // Assumption: all the conflicting attributes between left and right have been resolved
+          val newChild = child transformUp {
+            case p: Project =>
+              p.copy(projectList = p.projectList ++
+                findNotResolvedMissingAttrs(p.outputSet, p.inputSet, missingResolvableAttrs))
+            case w: Window =>
+              w.copy(projectList = w.projectList ++
+                findNotResolvedMissingAttrs(w.outputSet, w.inputSet, missingResolvableAttrs))
+            case a: Aggregate =>
+              val newAggregateExpressions = a.aggregateExpressions ++
+                findNotResolvedMissingAttrs(
+                  a.aggregateExpressions, a.inputSet, missingResolvableAttrs)
+              a.copy(aggregateExpressions = newAggregateExpressions)
+            case o => o
+          }
+
+          // Add missing attributes and then project them away after the sort.
+          Project(child.output,
+            Sort(newOrdering, s.global, newChild))
+        }
+    }
+
+    private def findNotResolvedMissingAttrs(
+        outputSet: AttributeSet,
+        inputSet: AttributeSet,
+        missingAttrs: Seq[Attribute]): Seq[Attribute] = {
+      val resolvedAttrs =
+        missingAttrs.filter(attr => inputSet.exists(_.semanticEquals(attr)))
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
```

Comment:
If we do not resolve sort references whose sources are below `subquery`, we can do it in your way. Otherwise, `semanticEquals` has to be used. Thanks! Let me use the simplified way.